x-pack/plugins/apm/server/lib/helpers/create_es_client/cancel_es_request_on_abort.ts

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { TransportRequestPromise } from '@elastic/elasticsearch/lib/Transport';
+import { KibanaRequest } from 'src/core/server';
+
+export function cancelEsRequestOnAbort<T extends TransportRequestPromise<any>>(
+  promise: T,
+  request: KibanaRequest
+) {
+  const subscription = request.events.aborted$.subscribe(() => {
+    promise.abort();
+  });
+
+  promise.then(
+    () => subscription.unsubscribe(),
+    () => subscription.unsubscribe()
```

Comment:
The second argument to `.then()` is a failure callback. The difference between `.then(onSuccess, onFailure)` and `.then(onSuccess).catch(onFailure)` is that the former calls onFailure only when the original promise is rejected, the second is also called when the `onSuccess` handler throws. I will add a comment.

Reply:
Isn't this what [`finally`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally) is for (Available since Node 10). 

```ts
promise.finally(() => subscription.unsubscribe())
```

Reply:
Yes, it's what I used initially but it failed on CI (with the API tests). Admittedly I was too lazy to figure out why that happened 😀 I'll have another look.

Reply:
Oh duh I knew that I was just reading it as two chained `then`s for some reason. Thanks.

Reply:
Yeah, .finally() doesn't work on CI: https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/99947/execution/node/595/log/

Reply:
bummer. Okay, thanks for trying.